components/support/nimbus-fml/src/intermediate_representation.rs

```diff
@@ -46,6 +47,54 @@ pub enum TypeRef {
     Option(Box<TypeRef>),
 }
 
+/**
+ * An identifier derived from a `FilePath` of a top-level or importable FML file.
+ *
+ * An FML module is the conceptual FML file (and included FML files) that a single
+ * Kotlin or Swift file. It can be imported by other FML modules.
+ *
+ * It is somewhat distinct from the `FilePath` enum for three reasons:
+ *
+ * - a file path can specify a non-canonical representation of the path
+ * - a file path is difficult to serialize/deserialize
+ * - a module
+ *
+ */
+#[derive(PartialEq, Eq, PartialOrd, Ord, Debug, Hash, Clone, Serialize, Deserialize)]
+pub enum ModuleId {
+    Local(String),
+    Remote(String),
+}
+
+impl Default for ModuleId {
+    fn default() -> Self {
+        Self::Local("none".to_string())
+    }
+}
+
+impl TryFrom<&FilePath> for ModuleId {
+    type Error = FMLError;
+    fn try_from(path: &FilePath) -> Result<Self> {
+        Ok(match path {
+            FilePath::Local(p) => {
+                let p = p.canonicalize().map_err(|e| {
+                    FMLError::InvalidPath(format!("{}: {}", e, p.as_path().display()))
```

Comment:
I added a comment of why I did this: it's because this is the first place that gets to find that the user or the file has given a file path that's missing.